src/Xamarin.Android.Build.Tasks/Utilities/MavenExtensions.cs

```diff
@@ -0,0 +1,229 @@
+
+		var pom = method.Invoke (null, new [] { sr }) as Project;
+
+		return pom!;
+	}
+
+	public static Artifact? CheckForNeededParentPom (string pomFile)
+		=> ParsePom (pomFile).GetParentPom ();
+
+	public static Artifact? GetParentPom (this Project? pom)
+	{
+		if (pom?.Parent != null)
+			return new Artifact (pom.Parent.ArtifactId, pom.Parent.GroupId, pom.Parent.Version);
+
+		return null;
+	}
+
+	// Returns artifact output path
+	public static async Task<string?> DownloadPayload (Artifact artifact, string cacheDir, TaskLoggingHelper log)
+	{
+		var version = artifact.Versions.First ();
+
+		var output_directory = Path.Combine (cacheDir, artifact.GetRepositoryCacheName (), artifact.GroupId, artifact.Id, version);
+
+		Directory.CreateDirectory (output_directory);
+
+		var filename = $"{artifact.GroupId}_{artifact.Id}";
+		var jar_filename = Path.Combine (output_directory, Path.Combine ($"{filename}.jar"));
+		var aar_filename = Path.Combine (output_directory, Path.Combine ($"{filename}.aar"));
+
+		// We don't need to redownload if we already have a cached copy
+		if (File.Exists (jar_filename))
+			return jar_filename;
+
+		if (File.Exists (aar_filename))
+			return aar_filename;
+
+		if (await TryDownloadPayload (artifact, jar_filename) is not string jar_error)
+			return jar_filename;
+
+		if (await TryDownloadPayload (artifact, aar_filename) is not string aar_error)
+			return aar_filename;
+
+		log.LogError ("Cannot download artifact '{0}:{1}'.\n- {2}: {3}\n- {4}: {5}", artifact.GroupId, artifact.Id, Path.GetFileName (jar_filename), jar_error, Path.GetFileName (aar_filename), aar_error);
+
+		return null;
+	}
+
+	// Returns artifact output path
+	public static async Task<string?> DownloadPom (Artifact artifact, string cacheDir, TaskLoggingHelper log, bool isParent = false)
+	{
+		var version = artifact.Versions.First ();
+		var output_directory = Path.Combine (cacheDir, artifact.GetRepositoryCacheName (), artifact.GroupId, artifact.Id, version);
+
+		Directory.CreateDirectory (output_directory);
+
+		var filename = $"{artifact.GroupId}_{artifact.Id}";
+		var pom_filename = Path.Combine (output_directory, Path.Combine ($"{filename}.pom"));
+
+		// We don't need to redownload if we already have a cached copy
+		if (File.Exists (pom_filename))
+			return pom_filename;
+
+		if (await TryDownloadPayload (artifact, pom_filename) is not string pom_error)
+			return pom_filename;
+
+		log.LogError ("Cannot download {4}POM file for artifact '{0}:{1}'.\n- {2}: {3}", artifact.GroupId, artifact.Id, Path.GetFileName (pom_filename), pom_error, isParent ? "parent " : "");
+
+		return null;
+	}
+
+	// Return value indicates download success
+	static async Task<string?> TryDownloadPayload (Artifact artifact, string filename)
+	{
+		try {
+			using var src = await artifact.OpenLibraryFile (artifact.Versions.First (), Path.GetExtension (filename));
+			using var sw = File.Create (filename);
+
+			await src.CopyToAsync (sw);
```

Comment:
Overload for `CancellationToken` should exist for this one.

Reply:
👍  Fixed!